src/uart.rs

```diff
@@ -0,0 +1,380 @@
+//! Access to [RIOT's UART](https://doc.riot-os.org/group__drivers__periph__uart.html)
+//!
+//! Author: Kilian Barning <barning@uni-bremen.de>
+#![allow(dead_code)]
+
+use core::marker::PhantomData;
+use core::ptr;
+
+use crate::error::{NegativeErrorExt, NumericError};
+use riot_sys::libc::{c_uint, c_void};
+use riot_sys::*;
+
+/// This enum representatives the status returned by various `UART`-functions
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum UartDeviceError {
+    InvalidDevice,
+    UnsupportedConfig,
+    Other,
+}
+
+impl UartDeviceError {
+    /// Converts the given `c_int` into the matching Enum representation
+    fn from_c(n: NumericError) -> Self {
+        const _ENODEV: isize = -(ENODEV as isize);
+        const _ENOTSUP: isize = -(ENOTSUP as isize);
+        match n.number {
+            _ENODEV => Self::InvalidDevice,
+            _ENOTSUP => Self::UnsupportedConfig,
+            _ => Self::Other,
+        }
+    }
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum DataBits {
+    Five,
+    Six,
+    Seven,
+    Eight,
+}
+
+impl DataBits {
+    fn to_c(self) -> uart_data_bits_t {
+        match self {
+            Self::Five => uart_data_bits_t_UART_DATA_BITS_5,
+            Self::Six => uart_data_bits_t_UART_DATA_BITS_6,
+            Self::Seven => uart_data_bits_t_UART_DATA_BITS_7,
+            Self::Eight => uart_data_bits_t_UART_DATA_BITS_8,
+        }
+    }
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Parity {
+    None,
+    Even,
+    Odd,
+    Mark,
+    Space,
+}
+
+impl Parity {
+    fn to_c(self) -> uart_parity_t {
+        match self {
+            Self::None => uart_parity_t_UART_PARITY_NONE,
+            Self::Even => uart_parity_t_UART_PARITY_EVEN,
+            Self::Odd => uart_parity_t_UART_PARITY_ODD,
+            Self::Mark => uart_parity_t_UART_PARITY_MARK,
+            Self::Space => uart_parity_t_UART_PARITY_SPACE,
+        }
+    }
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum StopBits {
+    One,
+    Two,
+}
+
+impl StopBits {
+    fn to_c(self) -> uart_stop_bits_t {
+        match self {
+            Self::One => uart_stop_bits_t_UART_STOP_BITS_1,
+            Self::Two => uart_stop_bits_t_UART_STOP_BITS_2,
+        }
+    }
+}
+
+/// This struct contains the `UART` device and handles all operation regarding it
+///
+/// [UART implementation]: https://doc.riot-os.org/group__drivers__periph__uart.html
+#[derive(Debug)]
+pub struct UartDevice<'scope> {
+    dev: uart_t,
+    _scope: PhantomData<&'scope ()>,
+}
+
+impl<'scope> UartDevice<'scope> {
+    fn construct_uart<F>(
+        index: usize,
+        baud: u32,
+        user_callback: &'scope mut F,
+    ) -> Result<Self, UartDeviceError>
+    where
+        F: FnMut(u8) + Sync + 'scope,
```

Comment:
```suggestion
        F: FnMut(u8) + Send + 'scope,
```

I might have suggested Sync earlier, but Sync is about shared references, and we're passing the exclusive reference to a thread here, thus moving the type.